src/main/java/org/spongepowered/common/mixin/core/network/MixinNetHandlerPlayServer.java

```diff
@@ -421,6 +426,34 @@ public void processCreativeInventoryAction(CPacketCreativeInventoryAction packet
         }
     }
 
+    /**
+     * @auther karayuu - June 21st, 2019 - MineCraft ver1.12.2
+     * @reason When a player use the spectator menu to teleport anyone,
+     * Sponge should fire the MoveEntityEvent.Teleport.
+     *
+     * @param packet The spectator packet
+     * @param entity The entity to teleport to
+     */
+    @Inject(
+            method = "handleSpectate",
+            at = @At(
+                    value = "FIELD",
+                    target = "Lnet/minecraft/entity/Entity;world:Lnet/minecraft/world/World;",
+                    ordinal = 0
+            ),
+            locals = LocalCapture.CAPTURE_FAILHARD,
+            cancellable = true
+    )
+    private void onHandleSpectateForSponge(CPacketSpectate packet, CallbackInfo ci, Entity entity) {
+        Transform<World> fromTransform = ((IMixinEntityPlayerMP) this.player).getTransform();
+        Transform<World> toTransform = ((IMixinEntity) entity).getTransform();
+
+        MoveEntityEvent.Teleport event = EntityUtil.handleDisplaceEntityTeleportEvent(this.player, fromTransform, toTransform);
```

Comment:
What happens if a plug-in changes the value of `event.getToTransform()`?

Reply:
Oh sorry.
Currently, even though Transform is changed by a plug-in, the player's transform doesn't change.
The change doesn't apply to NetHandlerPlayServer.
I'll fix this.